taichi/ir/snode.cpp

```diff
@@ -25,11 +25,13 @@ SNode &SNode::insert_children(SNodeType t) {
   TI_ASSERT(t != SNodeType::root);
 
   auto new_ch = std::make_unique<SNode>(depth + 1, t);
-  new_ch->is_path_all_dense = (is_path_all_dense && ((t == SNodeType::dense) ||
-                                                     (t == SNodeType::place)));
+  new_ch->is_path_all_dense = (is_path_all_dense && !new_ch->need_activation());
```

Comment:
In Taichi we demote struct-fors with all SNodes on the path being non-sparse into simple range-fors. This PR improves the coverage of "non-sparse".